### PR TITLE
[Accelerator 4] Adding Event creation, polling and subscription implementation

### PR DESCRIPTION
**Pull request by Ashi1993**
## [Accelerator 4] Adding Event creation, polling and subscription implementation

> Explain  in a few lines the purpose of this pull request

**Issue link:** *required*

**Doc Issue:** *Optional, link issue from [documentation repository](https://github.com/wso2/docs-ob/issues)*

**Applicable Labels:** *Spec, product, version, type (specify requested labels)*

------

### Development Checklist

1. [ ] Build complete solution with pull request in place.
2. [ ] Ran checkstyle plugin with pull request in place.
3. [ ] Ran Findbugs plugin with pull request in place.
4. [ ] Ran FindSecurityBugs plugin and verified report.
5. [ ] Formatted code according to WSO2 code style.
6. [ ] Have you verified the PR doesn't commit any keys, passwords, tokens, usernames, or other secrets?
7. [ ] Migration scripts written (if applicable).
8. [ ] Have you followed secure coding standards in [WSO2 Secure Engineering Guidelines](http://wso2.com/technical-reports/wso2-secure-engineering-guidelines)?

### Testing Checklist

1. [ ] Written unit tests.
2. [ ] Verified tests in multiple database environments (if applicable).
3. [ ] Tested with BI enabled  (if applicable).

### CODE REVIEW

File: financial-services-accelerator/accelerators/fs-is/carbon-home/repository/conf/financial-services.xml [hunk 1]
```diff
@@ -82,4 +82,31 @@
...
+            <NotificationGenerator>org.wso2.financial.services.accelerator.event.notifications.service.DefaultEventNotificationGenerator</NotificationGenerator>
+            <TokenIssuer>www.wso2.com</TokenIssuer>
+            <NumberOfSetsToReturn>5</NumberOfSetsToReturn>
+        </NotificationGeneration>
+        <EventCreationHandler>org.wso2.financial.services.accelerator.event.notifications.service.handler.DefaultEventCreationServiceHandler</EventCreationHandler>
+        <EventPollingHandler>org.wso2.financial.services.accelerator.event.notifications.service.handler.DefaultEventPollingServiceHandler</EventPollingHandler>
+        <EventSubscriptionHandler>org.wso2.financial.services.accelerator.event.notifications.service.handler.DefaultEventSubscriptionServiceHandler</EventSubscriptionHandler>
+        <PollingResponseParams>
+            <IsSubClaimAvailable>true</IsSubClaimAvailable>
+            <IsTxnClaimAvailable>{{financial_services.event.notifications.set_txn_claim_included}}</IsTxnClaimAvailable>
```
Comment (RivinduM) [inline financial-services-accelerator/accelerators/fs-is/carbon-home/repository/conf/financial-services.xml:96]: Is this a mistake?

Comment (Ashi1993) [reply financial-services-accelerator/accelerators/fs-is/carbon-home/repository/conf/financial-services.xml]: Fixed in https://github.com/wso2/financial-services-accelerator/pull/159/commits/247fe15ce5007a468941c9ad2df3af6a7185c0fe

**Review by RivinduM — APPROVED** [review]

File: financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/DefaultEventNotificationGenerator.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        return notificationResponse;
+    }
+
+    @Generated(message = "Excluded from tests as using a util method from a different package")
+    public String generateEventNotification(JsonNode jsonNode) throws FSEventNotificationException {
+
+        String payload = EventNotificationServiceUtil.getCustomNotificationPayload(jsonNode);
+        try {
+            return JWTUtils.signJWTWithDefaultKey(payload);
+        } catch (Exception e) {
```
Comment (imesh94) [inline financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/DefaultEventNotificationGenerator.java:73]: Need to catch a specific exception

File: financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/util/EventNotificationServiceUtil.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+    /**
+     * Method to modify event notification payload with custom eventValues.
+     *
+     * @param jsonNode Json Node to convert
+     * @return String eventNotificationPayload
+     */
+    public static String getCustomNotificationPayload(JsonNode jsonNode) {
+
+        return jsonNode.toString();
```
Comment (imesh94) [inline financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/util/EventNotificationServiceUtil.java:83]: Can't we do a .toString() at the caller without using a new method?

File: financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/util/EventNotificationServiceUtil.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+    /**
+     * Get the callback URL of the TPP from the Subscription Object.
+     *
+     * @param clientID client ID of the TPP
+     * @return callback URL of the TPP
+     */
+    public static String getCallbackURL(String clientID) {
+
+        return "http://localhost:8080/sample-tpp-server";
```
Comment (imesh94) [inline financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/util/EventNotificationServiceUtil.java:140]: Why are we returning a hardcoded url here?

File: financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/DefaultEventNotificationGenerator.java (hunk 2 above)
Comment (Ashi1993) [reply financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/DefaultEventNotificationGenerator.java:73]: To get the default key of the server, I am using keyStoreManager.getDefaultPrivateKey() in IS and that method throws an Exception. That's why I am catching that here

File: financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/util/EventNotificationServiceUtil.java (hunk 4 above)
Comment (Ashi1993) [reply financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/util/EventNotificationServiceUtil.java:140]: This is fixed from https://github.com/wso2/financial-services-accelerator/pull/193

File: financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/handler/DefaultEventSubscriptionServiceHandler.java [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+                log.error("Event subscription not found");
+                throw new FSEventNotificationException(HttpStatus.SC_BAD_REQUEST, "Event subscription not found");
+            }
+
+            EventSubscriptionResponse eventSubscriptionResponse = new EventSubscriptionResponse();
+            eventSubscriptionResponse.setResponseStatus(HttpStatus.SC_NO_CONTENT);
+            return eventSubscriptionResponse;
+        } catch (FSEventNotificationException e) {
+            log.error("Error occurred while deleting event subscription", e);
+            throw new FSEventNotificationException(HttpStatus.SC_BAD_REQUEST, e.getMessage(), e);
```
[the remark this answers is not in the document]
Comment (Ashi1993) [reply financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/handler/DefaultEventSubscriptionServiceHandler.java]: Fixed in https://github.com/wso2/financial-services-accelerator/pull/159/commits/782acc3a382f5f5f2905b4f3d07b83526261b666

File: financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/util/EventNotificationServiceUtil.java (hunk 3 above)
Comment (Ashi1993) [reply financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/util/EventNotificationServiceUtil.java]: Fixed in https://github.com/wso2/financial-services-accelerator/pull/159/commits/782acc3a382f5f5f2905b4f3d07b83526261b666

File: financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/internal/EventNotificationComponent.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+ */
+@Component(
+    name = "org.wso2.financial.services.accelerator.event.notifications.service.internal.EventNotificationComponent",
+    immediate = true)
+public class EventNotificationComponent {
+    private static Log log = LogFactory.getLog(EventNotificationComponent.class);
+
+    @Activate
+    protected void activate(ComponentContext context) {
+        if (log.isDebugEnabled()) {
```
Comment (imesh94) [inline financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/internal/EventNotificationComponent.java:44]: if condition can be removed

Comment (Ashi1993) [reply financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/internal/EventNotificationComponent.java]: FIxed in https://github.com/wso2/financial-services-accelerator/pull/159/commits/9fdacbc16042a3fdf8499672b5b76a49be0c68b2

File: financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/EventPollingService.java [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+                DatabaseUtils.commitTransaction(connection);
+
+                return aggregatedPollingResponse;
+            } catch (FSEventNotificationException |
+                     IOException | JOSEException | IdentityOAuth2Exception e) {
+                log.debug("Error when retrieving FS event notifications.", e);
+                DatabaseUtils.rollbackTransaction(connection);
+                throw new FSEventNotificationException("Error when retrieving FS event notifications.", e);
+            } finally {
+                log.debug(EventNotificationConstants.DATABASE_CONNECTION_CLOSE_LOG_MSG);
```
Comment (imesh94) [inline financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/EventPollingService.java:154]: Some classes are missing this debug log. Shall we make it consistent?

Comment (Ashi1993) [reply financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/EventPollingService.java:154]: Fixed in https://github.com/wso2/financial-services-accelerator/pull/159/commits/4e3160f15e06259be81743872bbf3ecd1c33db7e

**Review by imesh94 — APPROVED** [review]

**Pull request merged**